controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java

```diff
@@ -326,9 +326,18 @@ public ActivateResult deploy(ApplicationId applicationId, ZoneId zone,
                         ? triggered.sourceApplication().orElse(triggered.application())
                         : triggered.application();
 
-                applicationPackage = application.get().deploymentJobs().deployedInternally()
-                        ? new ApplicationPackage(applicationStore.getApplicationPackage(application.get().id(), applicationVersion.id()))
-                        : new ApplicationPackage(artifactRepository.getApplicationPackage(application.get().id(), applicationVersion.id()));
+                try {
+                    applicationPackage = application.get().deploymentJobs().deployedInternally()
+                            ? new ApplicationPackage(applicationStore.getApplicationPackage(application.get().id(), applicationVersion.id()))
+                            : new ApplicationPackage(artifactRepository.getApplicationPackage(application.get().id(), applicationVersion.id()));
+                }
+                catch (RuntimeException e) { // If application has switched deployment pipeline, artifacts stored prior to the switch are in the other artifact store.
```

Comment:
The artefacts are stored with their build number and the 9 first characters of their commit hash as keys, so that won't happen; the other artefact store will just throw an exception as well.

Reply:
Ok

Reply:
If only we had (discipline enough to properly use) checked exceptions ...

Reply:
No they don't work at scale because everybody en route need to catch-and rethrow *correctly* to avoid abstraction leaks.

Reply:
You get abstraction leaks with runtime exceptions, too, whenever you need to catch something that doesn't belong on your abstraction level. And then you have happy hours ahead trying to catch the right thing. So it's less visible, and not necessarily a problem, but a bigger one when it is.

Reply:
Yes, but the compiler don't see it, and you don't need to either, most of the time, since you can catch a generic exception.

If most exceptions got handled by exception-specific recovery code it would be worth declaring them, but most of the time they are handled in catch-alls at the container level.

Reply:
The problem is that it's hard to know which exceptions will be meaningful to catch (and declare), because this depends on the context where the API is used. So you're all but concluding that an exception that necessitates a certain action should be declared ... yet all I hear is to not use checked exceptions?

Reply:
Yes, checked exceptions are fine when you know some special action is needed ... but that is rare and then it should probably not be an exception.

There's some grey area as always and imo it is better to err on the side of no checking.